### ISSUE TITLE
How to get running without port forwarding?

### ISSUE THREAD

**Issue by lukebrooker** [cjryan/znc-openshift-quickstart#1]
Hi,

Thanks for creating this. I was able to get everything running, but I still need port forwarding, which kind of defeats the purpose of using ZNC in the first place.

I've tried using the ports that OpenShift has open but they are being used already. Where you able to get it working without port forwarding? If so, how?

Thanks

**Comment by cjryan** [comment]
Hi @lukebrooker, thanks for taking a look at this! You're right about port forwarding, I seem remember getting this working without it; I'll spend some time today and try to do it again. I'll update the docs when I find out how. Thanks!

**Comment by cjryan** [comment]
@lukebrooker thanks to some late-night help from @csrwng, it seems that I need to create a new diy cart, and expose the endpoints there. I'll update the repo as soon as that happens, hopefully shortly. Thanks!

**Comment by rnc** [comment]
Hi! Saw the link to this on mojo ; have you managed to get it working without port forwarding ?

**Comment by whispy** [comment]
I'm running into an issue where my connection is timing out when trying to access ZNC via IRC client or via web browser. I'm wondering if this issue is related to this port forwarding problem. Any idea @cjryan ?

**Comment by cjryan** [comment]
@whispy thanks for taking a look into ZNC! Are the timeouts consistent? I.e. do they always happen after a set amount of time? Or do the timeout lengths vary? Also, is there an error message that appears? Lastly, are you noticing the timeout in the IRC client, or in the rhc cli client? Thanks!

**Comment by whispy** [comment]
Yes, it's whenever I try to connect to my ZNC gear on Openshift. Not sure about the length between the timeouts.

The error message is just a 'connection timed out' message. And it's in the IRC client.

**Comment by cjryan** [comment]
@whispy ok, thanks. Any ballpark ideas about how long it is timing out? A minute, an hr, etc?

**Comment by whispy** [comment]
Less than a minute, I'd say. If you're on freenode or another server, I can chat with you about it.

EDIT: And my ZNC process seems to be working, as I see it logged in to IRC alongside me and the process is running on the gear. I just can't access it.

**Comment by whispy** [comment]
Also, I tried the `rhc port-forward -a [appname]` command, but I get `Error trying to forward ports`.

**Comment by whispy** [comment]
The port forwarding issue I experienced was due to me using an SSH private key that was named something other than `id_rsa` (`opensh_id_rsa` in this case). Apparently, `rhc` uses two different SSH agents — the system one for some commands (e.g. `rhc ssh`) and NET::SSH for others (e.g. `rhc port-forward`). NET::SSH was trying to use the nonexistent `id_rsa` key, even though the one I wanted it to use was `opensh_id_rsa`. Renaming `opensh_id_rsa` to `id_rsa` fixed the issue and allowed me to port forward.

**Comment by niklasp** [comment]
I also can not get it working without port forwarding which would be really important to make this a functional cartridge

**Comment by cjryan** [comment]
@niklasp Thanks for taking a look at the cartridge, and for the feedback. I'll spend some time this week and try to get this operational. I encountered issues in the past with OpenShift port restrictions, hopefully I can modify a 'mock' cartridge to do the work for us. Thanks!

**Comment by rnc** [comment]
@cjryan Did you manage to make any progress on the port forwarding / ssl connections?

**Comment by cjryan** [comment]
All, apologies that this fix has taken so long; with the way that OpenShift v2 was set up, it was difficult to change the ports in the cartridge to get around the port forwarding issue. Additionally, v2 is stuck at an older version of GCC, which means we're stuck with an older version of ZNC.

I would like to port this application to OpenShift v3, which will make it much simpler to run. Once v3 goes live (which, should be shortly), I will port the app, and we will get around the port forwarding issue, as well as run the newest version of GCC/VNC in RHEL.

https://www.openshift.com/devpreview/

**Comment by rnc** [comment]
Now openshift v3 is live is it worth resurrecting this issue ? :-D
